File: client-android/app/src/main/java/stasis/client_android/api/clients/MockServerCoreEndpointClient.kt
package stasis.client_android.api.clients

import okio.Buffer
import okio.Source
import stasis.client_android.lib.api.clients.ServerCoreEndpointClient
import stasis.client_android.lib.compression.Compression
import stasis.client_android.lib.encryption.Aes
import stasis.client_android.lib.encryption.secrets.DeviceSecret
import stasis.client_android.lib.model.core.CrateId
import stasis.client_android.lib.model.core.Manifest
import stasis.client_android.lib.model.core.NodeId

class MockServerCoreEndpointClient(
    private val deviceSecret: () -> DeviceSecret
) : ServerCoreEndpointClient {
    override val self: NodeId = MockConfig.DeviceNode

    override val server: String = MockConfig.ServerCore

    override suspend fun push(manifest: Manifest, content: Source) = Unit

    override suspend fun pull(crate: CrateId): Source? {
        val sample = MockDatasetContent.byCrate[crate] ?: return null
        if (sample.content.isEmpty()) return null

        val fileSecret = deviceSecret().toFileSecret(sample.cratePath, sample.checksum)
        val compressed = Compression.fromString(sample.compression).compress(Buffer().write(sample.content))

        return Aes.encrypt(compressed, fileSecret)
    }
}

File: client-ios/app/StasisClient/Mocks/MockServerCoreEndpointClient.swift
#if DEBUG
import Foundation
import StasisClientLib

struct MockServerCoreEndpointClient: ServerCoreEndpointClient {
    let selfNode: NodeId = MockConfig.deviceNode
    let server: String = MockConfig.serverCore

    private let resolveSecret: (@Sendable () -> DeviceSecret)?
    private let crates: [CrateId: MockLibraryFixtures.Fixture]

    init() {
        self.resolveSecret = nil
        self.crates = [:]
    }

    init(resolveSecret: @escaping @Sendable () -> DeviceSecret, crates: [CrateId: MockLibraryFixtures.Fixture]) {
        self.resolveSecret = resolveSecret
        self.crates = crates
    }

    func push(manifest: Manifest, content: Data) async throws {}

    func pull(crate: CrateId) async throws -> Data? {
        guard let fixture = crates[crate], let resolveSecret else { return nil }
        let secret = resolveSecret().toFileSecret(forFile: fixture.crateKey, checksum: fixture.checksum)
        return try secret.encrypt(fixture.plaintext)
    }
}
#endif
